shell_automaton/src/protocol_runner/protocol_runner_state.rs

```diff
@@ -27,6 +29,10 @@ pub enum ProtocolRunnerState {
     #[from]
     Init(ProtocolRunnerInitState),
 
+    /// Load context's current head
+    #[from]
+    GetCurrentHead(ProtocolRunnerCurrentHeadState),
```

Comment:
@sebastiencs I was thinking, do we really need a new separate message for this or is it something that would make sense as part of the context initialization step? (not worth to change it now that it is working, but something to keep in mind for the future once we start cleaning up all the protocol-runner stuff)

Reply:
I did it in a separate message/step to keep initializing the protocol runner as fast as possible.
The initialization of the context doesn't take more than 1 second, however retrieving the latest context hashes takes at least 15 seconds.

Reply:
Uhhh, why does it take so long to obtain the context hashes? we have those in the index, which just needs to be scanned from the end, right?

Reply:
Currently, the commit index is available after computing the crc32 of all files, and after the deserialization of the strings and shapes:
https://github.com/tezedge/tezedge/blob/develop/tezos/context/src/kv_store/persistent.rs#L660

In the case of the in-memory repository (loading from snapshot), it also has to commit the full tree (loaded from disk) into the in-mem repository:
https://github.com/tezedge/tezedge/blob/develop/tezos/context/src/kv_store/in_memory.rs#L561

Though with some changes, we could make the context hashes available faster

Reply:
For the in-memory case we still want the context to be ready before continuing, so getting the context hash faster would probably not help much. Same for the persisted context, but I think that part about deserializing strings and shapes could be speed up right? currently everything is deserialized although most of it is probably not used (there are shapes that repeat a lot and many that are used only one, or a few times and never again, and we could avoid paying the cost of parsing those and keeping them in memory).

Reply:
> I think that part about deserializing strings and shapes could be speed up right? currently everything is deserialized although most of it is probably not used

hmm yes we could do that, but we would have to read the full tree to get the list of strings & shapes used.
That's something we can test yes, and compare with the current speed

Reply:
We kind of know which ones are going to be used more right? and those could be kept on their own index. The others get another index that is never loaded at startup and works as an LRU cache. Anwyay, that would be for a separate PR